crypto/crypto/src/noise.rs

```diff
@@ -0,0 +1,587 @@
+pub static PROTOCOL_NAME: once_cell::sync::Lazy<Vec<u8>> = once_cell::sync::Lazy::new(|| {
+    if cfg!(test) {
+        b"Noise_IK_25519_AESGCM_SHA256\0\0\0\0".to_vec()
+    } else {
+        // only hash the protocol name if it's greater than HASHLEN
+        hash(b"Noise_IK_25519_AES128GCM_SHA3-256")
+    }
+});
+
+//
+// IK:
+//  <- s
+//  ...
+//  -> e, es, s, ss
+//  <- e, ee, se
+//
+
+//
+// Initiator
+// ---------
+//
+
+/// A key holder structure used for both initiators and responders.
+pub struct NoiseConfig {
+    private_key: x25519::StaticSecret,
+    public_key: [u8; X25519_PUBLIC_KEY_LENGTH],
+}
+
+/// Refer to the Noise protocol framework specification in order to understand these fields.
+pub struct InitiatorHandshakeState {
+    /// rolling hash
+    h: Vec<u8>,
+    /// chaining key
+    ck: Vec<u8>,
+    /// ephemeral key
+    e: x25519::StaticSecret,
+}
+
+impl NoiseConfig {
+    /// A peer must create a NoiseConfig through this function before being able to connect with other peers.
+    pub fn new(private_key: x25519::StaticSecret) -> Self {
+        let public_key = x25519::PublicKey::from(&private_key);
+        Self {
+            private_key,
+            public_key: public_key.as_bytes().to_owned(),
+        }
+    }
+
+    /// An initiator can use this function to initiate a handshake with a known responder.
+    pub fn initiate_connection(
+        &self,
+        rng: &mut (impl rand::RngCore + rand::CryptoRng),
+        prologue: &[u8],
+        remote_public: &x25519::PublicKey,
+        payload: Option<&[u8]>,
+    ) -> Result<(InitiatorHandshakeState, Vec<u8>), NoiseError> {
+        // checks
+        if let Some(payload) = payload {
+            if payload.len()
+                > MAX_SIZE_NOISE_MSG - 2 * X25519_PUBLIC_KEY_LENGTH - 2 * AES_GCM_TAGLEN
+            {
+                return Err(NoiseError::PayloadTooLarge);
+            }
+        }
+
+        // initialize
+        let mut h = PROTOCOL_NAME.clone();
+        let mut ck = PROTOCOL_NAME.clone();
+        mix_hash(&mut h, &prologue);
+        mix_hash(&mut h, remote_public.as_bytes());
+
+        // -> e
+        let e = if cfg!(test) {
+            let mut ephemeral_private = [0u8; X25519_PUBLIC_KEY_LENGTH];
+            rng.fill_bytes(&mut ephemeral_private);
+            x25519::StaticSecret::from(ephemeral_private)
+        } else {
+            x25519::StaticSecret::new(rng)
+        };
+        let e_pub = x25519::PublicKey::from(&e);
+        mix_hash(&mut h, e_pub.as_bytes());
+        let mut msg = e_pub.as_bytes().to_vec();
+
+        // -> es
+        let dh_output = e.diffie_hellman(remote_public);
+        let k = mix_key(&mut ck, dh_output.as_bytes())?;
+
+        // -> s
+        #[cfg(test)]
+        let aead = Aes256Gcm::new(GenericArray::clone_from_slice(&k));
+        #[cfg(not(test))]
+        let aead = Aes128Gcm::new(GenericArray::clone_from_slice(&k));
+
+        let msg_and_ad = Payload {
```

Comment:
msg_and_aad?

Reply:
same thing, I find aad redundant in general (and doesn't follow the ad in AEAD)